pkg/client/connection.go

```diff
@@ -30,4 +30,8 @@ type Connection interface {
 	// Close closes the connection, releasing all the resources that it uses. Once closed the
 	// connection can't be reused.
 	Close() error
+
+	// Publish / Subscribe interface
+	Publish(m Message, topic string) error
```

Comment:
how will you publish to a queue?
Sinc `Topic` and `Queue` are concepts driven by the queue definition (`anycast` vs `multicast`) I suggest we use the same API for both.
The publisher shouldn't care whether they publish to a queue or a topic.
So I suggest we use a generic name here like `destination`, `venue` or even `queue` which encapsulates the concept of `topic`

Reply:
Actually you don't _publish_ to a queue, rather you _send_ to a queue. So I think it is OK, at the level of abstraction of this library, to use the _publish_ and _topi_ concepts together.

In addition when publishing messages there are other parameters that are important, like the identifier of the client, if the messages should be persistent or not, etc. Because of that I think that we should have a separate `Publisher` interface, something like this:

```go
type Publisher interface {
    Publish(m Message error
}
```

The connection should then be able to create these publishers, supporting all the potential parameters (which may be added in the future). For example:

```go
publisher, err := connection.NewPublisherBuilder().
  Topic("mytopic").
  ClientID("myid").
  AnyOtherThingThatWeMayWantToAddInTheFuture().
  Build()
if err != nil {
   ...
}
```

This way we can add new settings to the publisher builder without breaking backwards compatibility.

Reply:
I don't see the advantage of using builders when GO allows initializing with string literals.
Builders require writing a lot of code of init/set functions we don't really need.

Reply:
Builders (or configuration structs, or list of options) allow you to add parameters without breaking backwards compatibility. Say for example that we have the `topic` parameter, and that it is used like this:

```go
publisher, err := connection.NewPublisher("mytopic)
```

Then you discover that you need to add a new `clientID` parameter. How would you add that parameter without forcing the user of the library to modify her code?

Reply:
I would probably add the clientID with a default value.

``` go
type Publisher struct {
  topic string
  clientID int
}

publisher := Publisher{
 topic: "mytopic",
 clientID: 123,
}
```

I'm not sure what you mean by breaking the code.
If the user wants to init this new param they will have to change their code anyway.
If they don't it will get a default value anyway, and the users code doesn't have to change.

Reply:
What you describe is a valid alternative. However it has the inconvenient that it makes the fields public, so the user can think that she can change them once the publisher is created, which may not be true. In addition there is no way to check if the parameters given by the user are valid or not, till you try to use them. For example, what if you want to verify that the topic does really exist, and return an error? To do so you need to call a method, not just create an struct. And to call a method you can use a builder:

```go
publisher, err := connection.NewPublisherBuilder().
  ParamA(...).
  ParamB(...).
  ...
  Build()
```

Or a configuration struct:

```go
publisher, err := connection.NewPublisher(&PublisherConfig{
  ParamA: ...,
  ParamB: ...,
  ...
})
```

I guess there are other possible mechanisms. But I think we need one of them; just creating the struct isn't enough.

Reply:
I think the convention in GO is to create a init function that returns a new instance of that type

``` go
func NewPublisher(conf PublisherConfig{...}) (*Publisher, error) 
```

In the Builder pattern, all the validations go in `Build()` method?